perceiver_pytorch/queries.py

```diff
@@ -0,0 +1,112 @@
+import torch
+from torch.distributions import uniform
+from typing import List, Union, Tuple
+from perceiver_pytorch.utils import encode_position
+from math import prod
+import einops
+import logging
+
+_LOG = logging.getLogger("perceiver.queries")
+_LOG.setLevel(logging.WARN)
+
+
+class LearnableQuery(torch.nn.Module):
```

Comment:
On the topic of position encoding for the queries...

I think the position encoding needs to be 'internally consistent' across each entire example.  That is, the position encoding should be consistent across data inputs _and_ the queries so the model can see that, say, the last timestep of the recent history in the input is immediately before the first timestep of the query.  And that, in general, the model can see that the timeseries of recent history and the timeseries of queries are two parts of a contiguous timeseries.

To give a concrete example: If the recent history in the input spans 11:00 to 11:55, and the query is for 12:00 to 12:55, then we want the model to see that the first timestep of the query is 5 minutes after the last timestep of the input... if that makes sense?!

One way to do this might be to encode the positions once for all timesteps in the example (i.e. the concatenation of the recent history and the forecast timesteps), and then concatenate the last `forecast_timesteps` of the position encoding to the queries?  Or something like that?!  Not sure what's best!

Perhaps this could wait for a future PR though!

This is just a hunch, of course!

And, ultimately, we might want the position encoding to include both the relative position ("the i<sup>th</sup> element in the array") and the absolute position in time and space ("2pm in South London").  But we can worry about that later :)

Reply:
Yeah, I might add that in a follow up PR, but it should be fairly easy to add those as options. I think it's a good idea to try at least!